packages/opentelemetry-cloud-trace-propagator/src/CloudPropagator.ts

```diff
@@ -0,0 +1,91 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  HttpTextPropagator,
+  Context,
+  SetterFunction,
+  GetterFunction,
+  TraceFlags,
+} from '@opentelemetry/api';
+import { decToHex, hexToDec } from 'hex2dec';
+import {
+  setExtractedSpanContext,
+  getParentSpanContext,
+} from '@opentelemetry/core';
+
+/**
+ * This file implements propagation for the Stackdriver Trace v1 Trace Context
+ * format.
+ *
+ * The header specification is:
+ * "X-Cloud-Trace-Context: TRACE_ID/SPAN_ID;o=TRACE_TRUE"
+ * Where:
+ *  {TRACE_ID} is a 32-character hexadecimal value representing a 128-bit
+ *    number. It should be unique between your requests, unless you
+ *    intentionally want to bundle the requests together.
+ *  {SPAN_ID} is the decimal representation of the (unsigned) span ID.
+ *    It should be randomly generated and unique in your trace. For subsequent requests,
+ *    set SPAN_ID to the span ID of the parent request.
+ *  {TRACE_TRUE} must be 1 to trace request. Specify 0 to not trace the request.
+ */
+
+/** Header that carries span context across Google infrastructure. */
+export const X_CLOUD_TRACE_HEADER = 'x-cloud-trace-context';
+
+export class CloudPropagator implements HttpTextPropagator {
+  inject(context: Context, carrier: unknown, setter: SetterFunction): void {
+    const spanContext = getParentSpanContext(context);
+    if (!spanContext) return;
+
+    const header = `${spanContext.traceId}/${hexToDec(
+      spanContext.spanId
+    )};o=${spanContext.traceFlags & TraceFlags.SAMPLED}`;
+
+    setter(carrier, X_CLOUD_TRACE_HEADER, header);
+  }
+
+  extract(context: Context, carrier: unknown, getter: GetterFunction): Context {
+    const traceContextHeader = getter(carrier, X_CLOUD_TRACE_HEADER);
+    const traceContextHeaderValue = Array.isArray(traceContextHeader)
+      ? traceContextHeader[0]
+      : traceContextHeader;
+    if (typeof traceContextHeaderValue !== 'string') {
+      return context;
+    }
+    const matches = traceContextHeaderValue.match(
+      /^([0-9a-fA-F]{32})(?:\/([0-9]+))(?:;o=(.*))?/
+    );
+
+    if (
+      !matches ||
+      matches[1] === '00000000000000000000000000000000' ||
+      matches[2] === '0000000000000000'
+    ) {
+      return context;
+    }
+
+    const spanContext = {
+      traceId: matches[1],
+      spanId: decToHex(matches[2], { prefix: false }).padStart(16, '0'),
+      traceFlags:
+        isNaN(Number(matches[3])) || matches[3] === '0'
```

Comment:
Hi, just had a quick question about usage of this feature when it's released. Let me know if there is a better place for this question.

I see this trace propagator decides whether to turn on tracing based on the header it receives.

In this case, how should the OpenTelemetry sampler be configured? If I only want my backend to sample traces that receive the `x-cloud-trace-context` header, would I use the [AlwaysOff](https://www.npmjs.com/package/@opentelemetry/core#alwaysoff) sampler? And this will override the AlwaysOff based on the header value?

Reply:
I believe the OpenTelemetry Gitter javascript channel is a good start for discussing the sampler aspects of this.
If you're a GCP customer, you could also try the support channels there.

A longer-term solution may be to add support for traceparent headers in ESP, of that's where your traffic is coming through.

Reply:
The Propagator only performs the injection and extraction of  a cross-cutting concern value as string key/values pairs into carriers that travel in-band across process boundaries.

Sampling (i.e. `sampler`) is resposible to control the noise and overhead introduced by OpenTelemetry by reducing the number of samples of traces collected and sent to the backend. 

> If I only want my backend to sample traces that receive the x-cloud-trace-context header, would I use the AlwaysOff sampler? And this will override the AlwaysOff based on the header value?

In this case I would suggest to use this (`CloudPropagator`) propagator along with either `AlwaysOn` or `Probability` built-in sampler.